proof/refine/ARM/Detype_R.thy

```diff
@@ -960,52 +1023,71 @@ lemma deleteObjects_corres:
+          apply (clarsimp simp: valid_mdb'_def valid_mdb_ctes_def valid_nullcaps_def)
+         apply clarsimp
+         apply (frule_tac cref="(aa, ba)" in cte_map_untyped_range,
+                erule cte_wp_at_weakenE[OF _ TrueI], assumption+)
          apply simp
-        apply (rule ext, clarsimp simp add: null_filter'_def
-                           map_to_ctes_delete[simplified field_simps])
-        apply (rule sym, rule ccontr, clarsimp)
-        apply (frule(2) pspace_relation_cte_wp_atI
-                        [OF state_relation_pspace_relation])
-        apply (elim exE)
-        apply (frule(4) cte_map_not_null_outside')
-         apply (rule cte_wp_at_weakenE, erule conjunct1)
-         apply (case_tac y, clarsimp)
-         apply (clarsimp simp: valid_mdb'_def valid_mdb_ctes_def
-                               valid_nullcaps_def)
-        apply clarsimp
-        apply (frule_tac cref="(aa, ba)" in cte_map_untyped_range,
-               erule cte_wp_at_weakenE[OF _ TrueI], assumption+)
-        apply simp
-       apply (rule detype_pspace_relation[simplified],
-              simp_all add: state_relation_pspace_relation valid_pspace_def)[1]
-        apply (simp add: valid_cap'_def capAligned_def)
-       apply (clarsimp simp: valid_cap_def, assumption)
-      apply (fastforce simp add: detype_def detype_ext_def intro!: ekheap_relation_detype)
-     apply (clarsimp simp: state_relation_def ghost_relation_of_heap
-                           detype_def)
+        apply (rule detype_pspace_relation[simplified],
+               simp_all add: state_relation_pspace_relation valid_pspace_def)[1]
+         apply (simp add: valid_cap'_def capAligned_def)
+        apply (clarsimp simp: valid_cap_def, assumption)
+       apply (fastforce simp add: detype_def detype_ext_def intro!: ekheap_relation_detype)
+      apply (rule detype_ready_queues_relation; blast?)
+       apply (clarsimp simp: deletionIsSafe_delete_locale_def)
+      apply (frule state_relation_ready_queues_relation)
+      apply (simp add: ready_queues_relation_def Let_def)
+     apply (clarsimp simp: state_relation_def ghost_relation_of_heap detype_def)
      apply (drule_tac t="gsUserPages s'" in sym)
      apply (drule_tac t="gsCNodes s'" in sym)
      apply (auto simp add: ups_of_heap_def cns_of_heap_def ext
-                 split: option.splits kernel_object.splits)[1]
+                    split: option.splits kernel_object.splits)[1]
     apply (simp add: valid_mdb_def)
    apply (wp hoare_vcg_ex_lift hoare_vcg_ball_lift | wps |
           simp add: invs_def valid_state_def valid_pspace_def
                     descendants_range_def | wp (once) hoare_drop_imps)+
   apply fastforce
   done
-
 end
 
 context delete_locale begin interpretation Arch . (*FIXME: arch_split*)
 
+lemma live_idle_untyped_range':
+  "\<lbrakk> ko_wp_at' P p s' \<or> p = idle_thread_ptr; \<And>ko. P ko \<Longrightarrow> live' ko \<rbrakk>
+   \<Longrightarrow> p \<notin> base_bits"
```

Comment:
This lemma looks a little bit strange to me. I feel like it would be more natural if the assumption was just in terms of `live'`, with any specific use cases that want to show a different `ko_wp_at'` condition then using `ko_wp_at'_weakenE` separately.

Reply:
I think this lemma is straight from `rt`. I actually thought it should have been in the separate PR I made for `Detype_R` but it's possible that it wasn't required for that PR and was required only for this PR. I will make this change, though